### PR TITLE
[Merged by Bors] - feat: the free locus of a (finitely presented) module

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/LocalizedModule/Basic.lean [hunk 1]
```diff
@@ -703,6 +703,40 @@ instance localizedModuleIsLocalizedModule :
         LocalizedModule.mk_cancel t]
   exists_of_eq eq1 := by simpa only [eq_comm, one_smul] using LocalizedModule.mk_eq.mp eq1
 
+
```
Comment (erdOne) [inline Mathlib/Algebra/Module/LocalizedModule/Basic.lean:706]:
```suggestion

```

**Comment by erdOne** [comment]
Regarding the large-import: `Mathlib.RingTheory.LocalProperties.Projective` is a leaf file that was only added on my last PR.

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 2]
```diff
@@ -25,8 +25,43 @@ import Mathlib.RingTheory.Localization.BaseChange
...
-variable [AddCommGroup N'] [Module R N'] (S : Submonoid R)
+universe uM
+
+variable {R N N' : Type*} {M : Type uM} [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N]
+variable [Module R N] [AddCommGroup N'] [Module R N'] (S : Submonoid R)
+
+theorem Module.free_of_isLocalizedModule {Rₛ Mₛ} [AddCommGroup Mₛ] [Module R Mₛ]
+    [CommRing Rₛ] [Algebra R Rₛ] [Module Rₛ Mₛ] [IsScalarTower R Rₛ Mₛ]
+    (S) (f : M →ₗ[R] Mₛ) [IsLocalization S Rₛ] [IsLocalizedModule S f] [Module.Free R M] :
+      Module.Free Rₛ Mₛ :=
```
Comment (jcommelin) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:36]:
```suggestion
    Module.Free Rₛ Mₛ :=
```

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 3]
```diff
@@ -25,8 +25,43 @@ import Mathlib.RingTheory.Localization.BaseChange
...
+/--
+Also see `IsLocalizedModule.lift_rank_eq` for a version for non-free modules,
+but requires `S` to not contain any zero-divisors.
+-/
+theorem Module.lift_rank_of_isLocalizedModule_of_free
+    (Rₛ : Type uR') {Mₛ : Type uM'} [AddCommGroup Mₛ] [Module R Mₛ]
+    [CommRing Rₛ] [Algebra R Rₛ] [Module Rₛ Mₛ] [IsScalarTower R Rₛ Mₛ] (S : Submonoid R)
+    (f : M →ₗ[R] Mₛ) [IsLocalization S Rₛ] [IsLocalizedModule S f] [Module.Free R M]
+    [Nontrivial Rₛ] :
+      Cardinal.lift.{uM} (Module.rank Rₛ Mₛ) = Cardinal.lift.{uM'} (Module.rank R M) := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:49]:
```suggestion
    Cardinal.lift.{uM} (Module.rank Rₛ Mₛ) = Cardinal.lift.{uM'} (Module.rank R M) := by
```

File: Mathlib/RingTheory/LocalProperties/Projective.lean [hunk 4]
```diff
@@ -25,8 +25,43 @@ import Mathlib.RingTheory.Localization.BaseChange
...
+    Cardinal.lift_one, one_mul, Cardinal.lift_lift] at this ⊢
+  convert this
+  exact Cardinal.lift_umax
+
+theorem Module.finrank_of_isLocalizedModule_of_free
+    (Rₛ : Type*) {Mₛ : Type*} [AddCommGroup Mₛ] [Module R Mₛ]
+    [CommRing Rₛ] [Algebra R Rₛ] [Module Rₛ Mₛ] [IsScalarTower R Rₛ Mₛ] (S : Submonoid R)
+    (f : M →ₗ[R] Mₛ) [IsLocalization S Rₛ] [IsLocalizedModule S f] [Module.Free R M]
+    [Nontrivial Rₛ] :
+      Module.finrank Rₛ Mₛ = Module.finrank R M := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/LocalProperties/Projective.lean:63]:
```suggestion
    Module.finrank Rₛ Mₛ = Module.finrank R M := by
```

File: Mathlib/Algebra/Module/FreeLocus.lean [hunk 5]
```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.PrimeSpectrum.Basic
```
Comment (jcommelin) [inline Mathlib/Algebra/Module/FreeLocus.lean:6]: Should this new file also be in the AG directory?

**Comment by jcommelin** [comment]
Very nice!

File: Mathlib/Algebra/Module/FreeLocus.lean (hunk 5 above)
Comment (erdOne) [reply Mathlib/Algebra/Module/FreeLocus.lean:6]: I'm not sure. Does `AlgebraicGeometry/CommutativeAlgebra/FreeLocus` make sense?

Comment (jcommelin) [reply Mathlib/Algebra/Module/FreeLocus.lean:6]: Yeah, I'm also not sure. Let's leave it as is, but make a mental note.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**